pycloudlib/azure/cloud.py

```diff
@@ -0,0 +1,851 @@
+
+        instance_info = {
+            "vm": vm,
+            "ip_address": ip_address_str,
+            "rg_name": self.resource_group.name
+        }
+
+        instance = AzureInstance(
+            client=self.compute_client,
+            key_pair=self.key_pair,
+            instance=instance_info
+        )
+
+        if wait:
+            instance.wait()
+
+        self.registered_instances[vm.name] = instance
+        return instance
+
+    def _create_ssh_resource(self, key_name):
+        """Create a ssh resource.
+
+        This method creates an Azure ssh resource to be associated
+        with a resource group.
+
+        Args:
+            key_name: string, The name of the ssh resource.
+
+        """
+        self.compute_client.ssh_public_keys.create(
+            self.resource_group.name,
+            key_name,
+            parameters={
+                "location": self.location,
+                "tags": {
+                    "name": self.tag
+                }
+            }
+        )
+
+    def create_key_pair(self, key_name):
+        """Create a pair of ssh keys.
+
+        This method creates an a pair of ssh keys in
+        the class resource group.
+
+        Args:
+            key_name: string, The name of the ssh resource.
+
+        """
+        self._create_ssh_resource(key_name)
+
+        ssh_call = self.compute_client.ssh_public_keys.generate_key_pair(
+            resource_group_name=self.resource_group.name,
+            ssh_public_key_name=key_name)
+
+        return ssh_call.public_key, ssh_call.private_key
+
+    def list_keys(self):
+        """List all ssh keys in the class resource group."""
+        ssh_public_keys = self.compute_client.ssh_public_keys
+
+        return [
+            ssh.name
+            for ssh in ssh_public_keys.list_by_resource_group(
+                self.resource_group.name)
+        ]
+
+    def delete_key(self, key_name):
+        """Delete a ssh key from the class resource group.
+
+        Args:
+            key_name: string, The name of the ssh resource.
+
+        """
+        ssh_public_keys = self.compute_client.ssh_public_keys
+        ssh_public_keys.delete(
+            resource_group_name=self.resource_group.name,
+            ssh_public_key_name=key_name
+        )
+
+    def use_key(self, public_key_path, private_key_path=None, name=None):
+        """Use an existing already uploaded key.
+
+        Args:
+            public_key_path: path to the public key to upload
+            private_key_path: path to the private key to upload
+            name: name to reference key by
+
+        """
+        if not name:
+            name = self.tag
+
+        self._log.debug('using SSH key %s', name)
+        self.key_pair = KeyPair(public_key_path, private_key_path, name)
+
+    def _list_instances(self):
+        """List instances in associated with a subscription id.
+
+        Returns:
+            An list of azure virtual machine associated with the subscription
```

Comment:
This is more lie _get_instances since it is not returning a list of instance names but actually the virtal machine object 
If we were making this a public method at some point, I'd like us to wrap each Virtual Machine as an AzureInstance.
```suggestion
    def _list_instances(self):
        """Return an iterable of Azure VirtualMachine instances associated with a subscription id.

        Returns:
            A list of azure virtual machine objects associated with the subscription
```

Reply:
Fair enough, I will update the name of the method. But to make each of the returned virtual machines an AzureInstance, we would need to retrieve the IP address of each virtual machine as well. Which can kind of expensive, considering we are now only using this method to retrieve the right instance.

But I agree that if we want to make this method public, it is a path we must follow to do that.